hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java

```diff
@@ -339,8 +339,7 @@ static <T extends CompoundPredicate> T flattenCompound(Predicate predicateLeft,
             Predicate[] left = getSubPredicatesIfClass(predicateLeft, klass);
             Predicate[] right = getSubPredicatesIfClass(predicateRight, klass);
 
-            predicates = new Predicate[left.length + right.length];
-            ArrayUtils.concat(left, right, predicates);
+            predicates = ArrayUtils.concat(left, right);
```

Comment:
Just to remind: We are adding here reflection cost

Reply:
> Just to remind: We are adding here reflection cost

I assumed that if were creating and discarding arrays here, that this wasn't a hot piece of code - but I've left the old method in place so we can switch this back if required.

Reply:
Please add a JMH benchmark.

And I'm not in favor of adding new methods if it isn't clear what kind of performance implication it has with respect to other methods.

Once the method is there, people will use it.

Reply:
Since there currently is only 1 place where this is used, do we really want a new utility method at all?

Reply:
> > Just to remind: We are adding here reflection cost
> 
> I _assumed_ that if were creating and discarding arrays here, that this wasn't a hot piece of code - but I've left the old method in place so we can switch this back if required.

You know what they say about assumption :) 

It is the mother of .... 

Also add this one on the JMH benchmark:

```
public static <T> T[] concat(T[] first, T... second) {
        T[] dst = (T[])new Object[first.length+second.length];
        concat(first, second, dst);
        return dst;
    }```

Reply:
> Please add a JMH benchmark.

https://github.com/hazelcast/hazelcast/pull/25953#issuecomment-1805537416

 
> And I'm not in favor of adding new methods if it isn't clear what kind of performance implication it has with respect to other methods.
> 
> Once the method is there, people will use it.

Good point - I've used a similar method in `commons-lang3` and I'd never considered how it's able to do this.


> Since there currently is only 1 place where this is used, do we really want a new utility method at all?

It's used in OS (x1), EE (x1) and my UCD branch (x1).

> Also add this one on the JMH benchmark:
> 
> ```
> public static <T> T[] concat(T[] first, T... second) {
>         T[] dst = (T[])new Object[first.length+second.length];
>         concat(first, second, dst);
>         return dst;
>     }```
> ```

That's a good idea, but unfortunately doesn't work and throws a `ClassCastException`.

Reply:
Maybe:

```java
public static <T> T[] concat(T[] first, T... second, Supplier<T[]> supplier) {
        T[] dst = (T[]) supplier.get();
        concat(first, second, dst);
        return dst;
    }
```

Then call as 
`int[] abc = concat(a, b, int[]::new);